core/client/src/children.rs

```diff
@@ -0,0 +1,123 @@
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+
+use std::collections::HashMap;
+use kvdb::{KeyValueDB, DBTransaction};
+use parity_codec::{Encode, Decode};
+use crate::error;
+use std::hash::Hash;
+
+
+/// Map of children blocks stored in memory for fast access.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ChildrenMap<K, V> where 
+	K: Eq + Hash + Clone + Encode + Decode,
+	V: Eq + Hash + Clone + Encode + Decode,
+{
+	storage: HashMap<K, Vec<V>>,
+}
+
+impl<K, V> ChildrenMap<K, V> where
+	K: Eq + Hash + Clone + Encode + Decode,
+	V: Eq + Hash + Clone + Encode + Decode,
+{
+	/// Creates an empty children map.
+	pub fn new() -> Self {
+		Self {
+			storage: HashMap::new(),
+		}
+	}
+
+	/// Returns the hashes of the children blocks of the block with `parent_hash`.
+	pub fn hashes(&self, db: &KeyValueDB, column: Option<u32>, prefix: &[u8], parent_hash: K) -> error::Result<Vec<V>> {
+		let mut buf = prefix.to_vec();
+		parent_hash.using_encoded(|s| buf.extend(s));
+
+		let raw_val_opt = match db.get(column, &buf[..]) {
+			Ok(raw_val_opt) => raw_val_opt,
+			Err(_) => return Err(error::ErrorKind::Backend("Error reading value from database".into()).into()),
+		};
+
+		let raw_val = match raw_val_opt {
+			Some(val) => val,
+			None => return Ok(Vec::new()),
+		};
+
+		let children: Vec<V> = match Decode::decode(&mut &raw_val[..]) {
+			Some(children) => children,
+			None => return Err(error::ErrorKind::Backend("Error decoding children".into()).into()),
+		};
+
+		Ok(children)
+	}
+
+	/// Returns the hashes of the children blocks of the block with `parent_hash`.
+	/// It doesn't read the database.
+	pub fn hashes_from_mem(&self, parent_hash: K) -> Vec<V> {
+		self.storage.get(&parent_hash).cloned().unwrap_or_default()
+	}
+
+	/// Import the hash `child_hash` as child of `parent_hash`.
+	/// It doesn't save changes to database.
+	pub fn import(&mut self, parent_hash: K, child_hash: V) {
+		self.storage.entry(parent_hash).or_insert_with(Vec::new).push(child_hash);
+	}
+
+	/// Prepare the transaction `tx` that saves the content of the ChildrenMap to database.
+	/// It clears the content of ChildrenMap.
+	pub fn prepare_transaction(&mut self, db: &KeyValueDB, tx: &mut DBTransaction, column: Option<u32>, prefix: &[u8])
+		-> error::Result<()> {
+		for (parent_hash, children) in self.storage.iter() {
```

Comment:
You should use `self.storage.drain(..)` here and drop `self.storage.clear()`. 
But the question is also, do we need the internal storage hashmap at all? 
As I have written above, you call `import` directly before calling `prepare_transaction`. As I see this, you could completely remove `import` and give the `parent_hash` and `child_hash` directly to this function as you will never have more than one child at all per `prepare_transaction`.

Reply:
I wonder the same. I may be confused, but apparently we need it here: https://github.com/paritytech/substrate/blob/343e54875a05c888b2ec775d71a2d5710863c4c0/core/client/src/in_mem.rs#L345:L374, since the description of `Blockchain` reads "In-memory blockchain" and is using the `leaves` module without writing to db. Otherwise, I would like to remove the hashmap since it creates inconsistency, either by removing this `children` method or replacing it with a db read.